### PR TITLE
New sample: Navigate route with rerouting

### PR DESCRIPTION
**Pull request by shubham7109**
## Description
PR to add a new Kotlin sample _"Navigate route with rerouting"_ in `Routing and Logistics` category.

## Links and Data
Sample Epic: `runtime/common-samples/issues/3511`

## What To Review
-  Review the code to make sure it is easy to follow like other samples on Android
- `README.md` and `README.metadata.json` files

## How to Test
Run the sample on the sample viewer or the repo.

## Screenshots
navigate-with-reroute

### CODE REVIEW

**Comment by shubham7109** [comment]
The data source is subject to change, we might use a portal item instead of a hard-coded JSON string

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+import com.esri.arcgisruntime.tasks.networkanalysis.RouteResult
+import com.esri.arcgisruntime.tasks.networkanalysis.RouteTask
+import com.esri.arcgisruntime.tasks.networkanalysis.Stop
+import java.util.Calendar
+import kotlin.math.roundToInt
+
+
+class MainActivity : AppCompatActivity() {
+
+    private val TAG = MainActivity::class.java.simpleName
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:56]: this is unused

File: kotlin/navigate-route-with-rerouting/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+3. Open your command prompt and navigate to the folder where you extracted the contents of the data from step 1.
+4. Execute the following commands:
+
+`adb push streetmap_SD.tpkx /Android/data/com.esri.arcgisruntime.sample.offlinerouting/files/san_diego/streetmap_SD.tpkx`
+
+`adb push sandiego.geodatabase /Android/data/com.esri.arcgisruntime.sample.offlinerouting/files/san_diego/sandiego.geodatabase`
+
+Link | Local Location
+---------|-------|
+|[San Diego streetmap TPKX](https://arcgisruntime.maps.arcgis.com/home/item.html?id=df193653ed39449195af0c9725701dca)| /Android/data/com.esri.arcgisruntime.sample.offlinerouting/files/streetmap_SD.tpkx |
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/README.md:58]: we're not using this it looks like (which I think is a good idea). Let's remove from README and metadata

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+                // destination location point: Fleet Science Center
+                val aerospaceMuseum = Point(-117.146679, 32.730351, SpatialReferences.getWgs84())
+                // create starting stop
+                val startingStop = Stop(conventionCenter).apply {
+                    name = "San Diego Convention Center"
+                }
+                // create destination stop
+                val destinationStop = Stop(aerospaceMuseum).apply {
+                    name = "RH Fleet Aerospace Museum"
+                }
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:194]:
```suggestion
                // starting location point: San Diego Convention Center
                val conventionCenter =
                    Point(-117.160386727, 32.706608, SpatialReferences.getWgs84())
                // create starting stop
                val startingStop = Stop(conventionCenter).apply {
                    name = "San Diego Convention Center"
                }
                // destination location point: Fleet Science Center
                val aerospaceMuseum = Point(-117.146679, 32.730351, SpatialReferences.getWgs84())
                // create destination stop
                val destinationStop = Stop(aerospaceMuseum).apply {
                    name = "RH Fleet Aerospace Museum"
                }
```

Comment (TADraeseke) [reply kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:194]: Better to keep the parameter close to where it's being added to the constructor

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+                    // enable navigation
+                    navigateButton.isEnabled = true
+                }
+            }
+        }
+    }
+
+    /**
+     * Starts the route navigation using a simulated datasource
+     * and sets up re-routing functionalities
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:236]:
```suggestion
     * and sets up re-routing functionalities.
```

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+    private val recenterButton: Button by lazy {
+        activityMainBinding.recenterButton
+    }
+
+    private val statusMessageTV: TextView by lazy {
+        activityMainBinding.statusMessageTV
+    }
+
+    // a list to keep track of directions solved by the route task
+    private val directionManeuvers = mutableListOf<DirectionManeuver>()
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:83]: I wonder if it's worth keeping a second list of the maneuvers. Couldn't we just reference route.directionManeuvers directly where we need it?

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+                mapView.locationDisplay.startAsync()
+                // add re-routing listeners
+                routeTracker?.addRerouteStartedListener(routeStartedListener)
+                routeTracker?.addRerouteCompletedListener(routeCompletedListener)
+            }
+        }
+    }
+
+    /**
+     * Removes the current route tracking status listener if new route loading has started
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:304]:
```suggestion
     * Removes the current route tracking status listener if new route loading has started.
```

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+     */
+    private val routeStartedListener: RouteTracker.RerouteStartedListener =
+        RouteTracker.RerouteStartedListener {
+            routeTracker?.removeTrackingStatusChangedListener(trackingStatusListener)
+        }
+
+    /**
+     * Removes the route tracking status listener if re-route loading has completed,
+     * sets up new direction maneuvers and adds a new route tracking listener
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:313]:
```suggestion
     * sets up new direction maneuvers and adds a new route tracking listener.
```

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+            directionManeuvers.clear()
+            // adds new direction maneuvers to the list
+            directionManeuvers.addAll(it.trackingStatus.routeResult.routes[0].directionManeuvers)
+            // re-add the event listeners for tracking status changes
+            routeTracker?.addTrackingStatusChangedListener(trackingStatusListener)
+        }
+
+
+    /**
+     * Calls updateTrackingStatus() on every route tacking status changed
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:328]:
```suggestion
     * Calls updateTrackingStatus() on every route tacking status changed.
```

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+     */
+    private val trackingStatusListener: RouteTracker.TrackingStatusChangedListener =
+        RouteTracker.TrackingStatusChangedListener {
+            updateTrackingStatus(it.trackingStatus)
+        }
+
+    /**
+     * Displays the route status descriptions and update route graphics as the
+     * the datasource updates on the map
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:337]:
```suggestion
     * the datasource updates on the map.
```

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+     */
+    private fun updateTrackingStatus(status: TrackingStatus) {
+        val statusMessage: StringBuilder = StringBuilder("")
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:340]:
```suggestion
        val statusMessage: StringBuilder = StringBuilder()
```

StringBuilder defaults to an empty string anyways I think

**Review by TADraeseke — CHANGES_REQUESTED** [review]
Great sample @shubham7109 -- tons of code here, but highly readable and consumable. Really nice work. Just a few small thoughts below 👍

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt (hunk 5 above)
Comment (shubham7109) [reply kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:83]: Removed `directionManeuvers` list 👍

**Comment by shubham7109** [comment]
Implemented PR feedback and added TTS Android speech engine

**Comment by shubham7109** [comment]
@TADraeseke Added a fix for the direction maneuvers, and the sample is now displaying the correct text.

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt (hunk 5 above)
Comment (shubham7109) [reply kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:88]: Turns out we need it haha.

File: kotlin/navigate-route-with-rerouting/README.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+While traveling between destinations, field workers use navigation to get live directions based on their locations. In cases where a field worker makes a wrong turn, or if the route suggested is blocked due to a road closure, it is necessary to calculate an alternate route to the original destination.
+
+## How to use the sample
+
+Tap "Navigate" to simulate traveling and to receive directions from a preset starting point to a preset destination. Observe how the route is recalculated when the simulation does not follow the suggested route. Tap "Recenter" to reposition the viewpoint. Tap "Reset" to start the simulation from the beginning.
+
+## How it works
+
+1. Create an `RouteTask` using local network data.
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/README.md:17]:
```suggestion
1. Create a `RouteTask` using local network data.
```

File: kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+        routeTask = RouteTask(
+            this,
+            getExternalFilesDir(null)?.path + "/sandiego.geodatabase",
+            "Streets_ND"
+        )
+
+        // load the route task
+        routeTask.loadAsync()
+        routeTask.addDoneLoadingListener {
+            if(routeTask.loadStatus == LoadStatus.LOADED){
```
Comment (TADraeseke) [inline kotlin/navigate-route-with-rerouting/src/main/java/com/esri/arcgisruntime/sample/navigateroutewithrerouting/MainActivity.kt:185]:
```suggestion
            if (routeTask.loadStatus == LoadStatus.LOADED){
```

Just need a Code -> Reformat code run I think

**Review by TADraeseke — APPROVED** [review]
LGTM! Nice job @shubham7109 -- just needs a code reformat.

**Review by saratchandrakarumuri — APPROVED** [review]
New sample is looking good

**Pull request merged**